src/mesh/coordinates.cxx

```diff
@@ -1695,11 +1715,20 @@ Coordinates::FieldMetric Coordinates::Grad2_par2(const Field2D& f, CELL_LOC outl
   TRACE("Coordinates::Grad2_par2( Field2D )");
   ASSERT1(location == outloc || (outloc == CELL_DEFAULT && location == f.getLocation()));
 
-  auto invSg = 1.0 / sqrt(g_22);
-  // Communicate to get parallel slices
-  localmesh->communicate(invSg);
-  auto result = DDY(invSg, outloc, method) * DDY(f, outloc, method) * invSg
-                + D2DY2(f, outloc, method) / g_22;
+  if (invSg == nullptr) {
+    invSg = std::make_unique<FieldMetric>();
+    (*invSg) = 1.0 / sqrt(g_22);
+  }
+  if (Grad2_par2_DDY_invSg == nullptr) {
```

Comment:
We presumably actually need a cached version for each possible `outloc` (and possibly `method` too?) -- unless `this` is already at `outloc`, in which case this happens automatically?